src/main/java/javax/validation/valueextraction/ValueExtractor.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Bean Validation API
+ *
+ * License: Apache License, Version 2.0
+ * See the license.txt file in the root directory or <http://www.apache.org/licenses/LICENSE-2.0>.
+ */
+package javax.validation.valueextraction;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import javax.validation.Path;
+
+/**
+ * Defines the logic used to extract the values to validate from an object of type {@code T}.
+ * <p>
+ * A {@link ValueExtractor} is tied to one specific type parameter of {@code T}. The {@link ExtractedValue} annotation
+ * is used to mark that type parameter.
+ * <p>
+ * The extracted values are passed to the corresponding method of the {@link ValueReceiver}.
+ *
+ * @author Gunnar Morling
+ * @author Guillaume Smet
+ */
+public interface ValueExtractor<T> {
+
+	/**
+	 * Extracts the values to validate from the original object.
+	 *
+	 * @param originalValue the original value from which we want to extract the values to validate
+	 * @param receiver the corresponding {@code ValueReceiver}
+	 */
+	void extractValues(T originalValue, ValueReceiver receiver);
+
+	/**
+	 * Provides a set of methods receiving value extracted by the {@link ValueExtractor}.
+	 * <p>
+	 * The value has to be passed to the method corresponding best to the type of the original value.
+	 */
+	interface ValueReceiver {
+
+		/**
+		 * Receives the value extracted from an object.
+		 *
+		 * @param object the value to validate
+		 * @param nodeName the name of the node to be added to the {@link Path}
+		 */
+		void value(Object object, String nodeName);
```

Comment:
@emmanuelbernard suggested - and I had the same inclination before - to switch the parameter order: first the "meta data", i.e. node name/key/index, and then the value. That seems more natural to use.